pymc/sampling/forward.py

```diff
@@ -497,22 +499,148 @@ def sample_posterior_predictive(
+
+    .. code:: python
+
+        from logging import getLogger
+        import pymc as pm
+
+        # Some environments like google colab suppress
+        # the default logging output of PyMC
+        getLogger("pymc").setLevel("INFO")
+
+        kwargs = {"progressbar": False, "random_seed": 0}
+
+        with pm.Model() as model:
+          x = pm.Normal("x")
+          y = pm.Normal("y")
+          z = pm.Normal("z", x + y**2)
+          det = pm.Deterministic("det", pm.math.exp(z))
+          obs = pm.Normal("obs", det, 1, observed=[20])
+
+          idata = pm.sample(tune=10, draws=10, chains=2, **kwargs)
+
+    Default behavior. Posterior predictive draws of ``obs``,
+    conditioned on the posterior draws of ``z`` found in the trace.
+
+    .. code:: python
+
+          pm.sample_posterior_predictive(idata, var_names=["obs"], **kwargs)
+          # Sampling: [obs]
+
+    Re-compute the deterministic variable ``det``,
+    conditioned on the posterior draws of ``z`` found in the trace.
+
+    .. code :: python
+
+          pm.sample_posterior_predictive(idata, var_names=["det"], **kwargs)
+          # Sampling: []
+
+    Posterior predictive draws of ``z`` and ``det``,
+    conditioned on the posterior draws of ``x``` and ``y``` found in the trace.
+
+    Note: It does not matter that ``z``` was not an observed variable.
+
+    .. code :: python
+
+          pm.sample_posterior_predictive(idata, var_names=["z", "det"], **kwargs)
+          # Sampling: [z]
+
+
+    Posterior predictive draws of ``y``, ``z`` and ``det``,
+    conditioned on the posterior draws of ``x`` found in the trace.
+
+    Note: The samples of ``y`` are equivalent to its prior, since it does not depend on any other variables.
+
+    .. code :: python
+
+          pm.sample_posterior_predictive(idata, var_names=["y", "z", "det"], **kwargs)
+          # Sampling: [y, z]
+
+
+    Same as before, except ``z`` is not stored in the returned trace.
+    For computing ``det`` we still have to sample ``z`` as it depends on ``y``, which is also being sampled.
+
+    .. code :: python
+
+          pm.sample_posterior_predictive(idata, var_names=["y", "det"], **kwargs)
+          # Sampling: [y, z]
+
+    Every random variable is sampled. This is equivalent to calling :func:~`pymc.sample_prior_predictive`
+
+    .. code :: python
+
+          pm.sample_posterior_predictive(idata, var_names=["x", "y", "z", "det", "obs"], **kwargs)
+          # Sampling: [x, y, z, obs]
+
+
+    Note that "sampling" a dependent deterministic does not force random variables to be sampled.
+
+    .. code :: python
+
+        with pm.Model() as model:
+          x = pm.Normal("x")
+          y = pm.Normal("y")
+          det_xy = pm.Deterministic("det_xy", x + y**2)
+          z = pm.Normal("z", x + y**2)
+          det_z = pm.Deterministic("det_z", pm.math.exp(z))
+          obs = pm.Normal("obs", det, 1, observed=[20])
```

Comment:
```suggestion
          obs = pm.Normal("obs", det_z, 1, observed=[20])
```

Reply:
Ah this is why it was failing, so it's actually working as expected